server/src/main/java/org/elasticsearch/action/admin/cluster/node/reload/NodesReloadSecureSettingsRequest.java

```diff
@@ -62,24 +68,60 @@ public ActionRequestValidationException validate() {
         return validationException;
     }
 
-    public String secureSettingsPassword() {
+    public SecureString secureSettingsPassword() {
         return secureSettingsPassword;
     }
 
-    public NodesReloadSecureSettingsRequest secureStorePassword(String secureStorePassword) {
+    public NodesReloadSecureSettingsRequest secureStorePassword(SecureString secureStorePassword) {
         this.secureSettingsPassword = secureStorePassword;
         return this;
     }
 
     @Override
     public void readFrom(StreamInput in) throws IOException {
         super.readFrom(in);
-        secureSettingsPassword = in.readString();
+        final byte[] passwordBytes = in.readByteArray();
+        try {
+            this.secureSettingsPassword = new SecureString(utf8BytesToChars(passwordBytes));
+        } finally {
+            Arrays.fill(passwordBytes, (byte) 0);
+        }
     }
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
-        out.writeString(secureSettingsPassword);
+        final byte[] passwordBytes = toUtf8Bytes(this.secureSettingsPassword.getChars());
+        try {
+            out.writeByteArray(passwordBytes);
+        } finally {
+            Arrays.fill(passwordBytes, (byte) 0);
+        }
+    }
+
+    /**
+     * Encodes the provided char[] to a UTF-8 byte[]. The provided char[] is not
+     * modified by this method, so the caller needs to take care of clearing the
+     * value if it is sensitive.
+     */
+    private static byte[] toUtf8Bytes(char[] chars) {
+        final CharBuffer charBuffer = CharBuffer.wrap(chars);
+        final ByteBuffer byteBuffer = StandardCharsets.UTF_8.encode(charBuffer);
+        final byte[] bytes = Arrays.copyOfRange(byteBuffer.array(), byteBuffer.position(), byteBuffer.limit());
+        Arrays.fill(byteBuffer.array(), (byte) 0); // clear sensitive data
+        return bytes;
+    }
+
+    /**
+     * Decodes the provided byte[] to a UTF-8 char[]. The provided byte[] is not
+     * modified by this method, so the caller needs to take care of clearing the
+     * value if it is sensitive.
+     */
+    public static char[] utf8BytesToChars(byte[] utf8Bytes) {
+        final ByteBuffer byteBuffer = ByteBuffer.wrap(utf8Bytes);
+        final CharBuffer charBuffer = StandardCharsets.UTF_8.decode(byteBuffer);
+        final char[] chars = Arrays.copyOfRange(charBuffer.array(), charBuffer.position(), charBuffer.limit());
+        Arrays.fill(charBuffer.array(), (char) 0); // clear sensitive data
+        return chars;
```

Comment:
Copied from: https://github.com/elastic/elasticsearch/blob/113c1916eeffb8fb453d56c97942eefa19bcb087/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/CharArrays.java#L18

See [previous comment](https://github.com/elastic/elasticsearch/pull/31261/files#r194657974) .

Some point worth raising here. In the original implementation (`CharArray#utf8BytesToChars`) the decoded `CharBuffer` is not cleared. It is `charBuffer.clear()` but should be `Arrays.fill(charBuffer.array(), (char) 0);`.
@jaymode Should I raise a PR to fix this?

Reply:
I think it is worth raising a separate PR to fix this and move the methods as I said before. I also think that this method needs to handle some other cases such as my suggestion above about the hasArray() and other items.